lib/TsconfigPathsPlugin.js

```diff
@@ -0,0 +1,238 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Natsu @xiaoxiaojx
+*/
+
+"use strict";
+
+const path = require("path");
+const { aliasResolveHandler } = require("./AliasUtils");
+const { modulesResolveHandler } = require("./ModulesUtils");
+
+/** @typedef {import("./Resolver")} Resolver */
+/** @typedef {{alias: string | string[] | false, name: string, onlyModule?: boolean}} AliasOption */
+/** @typedef {import("./Resolver").ResolveStepHook} ResolveStepHook */
+
+/**
+ * @typedef {object} TsconfigPathsPluginOptions
+ * @property {string=} configFile - Path to tsconfig.json file
+ * @property {string=} baseUrl - Override baseUrl from tsconfig.json
+ * @property {string=} context - Context directory for resolving relative paths
+ */
+
+module.exports = class TsconfigPathsPlugin {
+	/**
+	 * @param {string} configFile tsconfig file path
+	 */
+	constructor(configFile) {
+		this.configFile = configFile;
+	}
+
+	/**
+	 * @param {Resolver} resolver the resolver
+	 * @returns {void}
+	 */
+	apply(resolver) {
+		if (!this.configFile) return;
+
+		const aliasTarget = resolver.ensureHook("internal-resolve");
+		const moduleTarget = resolver.ensureHook("module");
+		const aliasOptionsPromise = this.loadAndConvertPaths(resolver);
+
+		resolver
+			.getHook("raw-resolve")
+			.tapAsync(
+				"TsconfigPathsPlugin",
+				async (request, resolveContext, callback) => {
+					try {
+						const aliasOptions = await aliasOptionsPromise;
+						aliasResolveHandler(
+							resolver,
+							aliasOptions.filter((option) => option.name !== "*"),
+							aliasTarget,
+							request,
+							resolveContext,
+							callback,
+						);
+					} catch (err) {
+						callback(/** @type {Error} */ (err));
+					}
+				},
+			);
+
+		resolver
+			.getHook("raw-module")
+			.tapAsync(
+				"TsconfigPathsPlugin",
+				async (request, resolveContext, callback) => {
+					try {
+						const aliasOptions = await aliasOptionsPromise;
+						const directories = aliasOptions
+							.filter((option) => option.name === "*")
+							.map((option) => /** @type {string} */ (option.alias));
+						if (directories.length === 0) return callback();
+
+						modulesResolveHandler(
+							resolver,
+							directories,
+							moduleTarget,
+							request,
+							resolveContext,
+							callback,
+						);
+					} catch (err) {
+						callback(/** @type {Error} */ (err));
+					}
+				},
+			);
+	}
+
+	/**
+	 * Load tsconfig.json (and referenced tsconfigs) and convert paths to AliasPlugin format
+	 * @param {Resolver} resolver the resolver
+	 * @returns {Promise<AliasOption[]>} Array of alias options for AliasPlugin
+	 */
+	async loadAndConvertPaths(resolver) {
+		try {
+			const fs = resolver.fileSystem;
+			const configPath = path.isAbsolute(this.configFile)
+				? this.configFile
+				: resolver.join(process.cwd(), this.configFile);
+
+			const mainOptions = await this.readTsconfigCompilerOptions(
```

Comment:
I think we can cache `readTsconfigCompilerOptions` using (and the same for `readTsconfigCompilerOptions` below)
```js
this.tsconfigProcessorCache = new WeakMap<TS_CONFIG_JSON_VALUE, { aliases: Aliases, references: References }>;
```

So we will read everything only on the first run, and will keep results until GC allow it

Feel free to think how we can avoid more tasks on the second and third calls

Reply:
the current implementation only reads tsconfig.json once during instantiation, so it already acts as a cache.
```js
apply(resolver) {
		 // 👇 it only reads once when the plugin is instantiated
		// similar to how this.tsconfigProcessorCache works.
		const aliasOptionsPromise = this.loadAndConvertPaths(resolver);

```
In most cases, TsconfigPathsPlugin will only be instantiated once, so aliasOptionsPromise acts as a cache that ensures it’s only read once.
In less common cases where the user calls ResolverFactory.create for every resolve, the plugin will be instantiated multiple times. In that case, if users want caching, they can enable it via the unsafeCache option.

Reply:
What if we do it like package.json and attach the data to [request.descriptionFileData](https://github.com/webpack/enhanced-resolve/blob/main/lib/DescriptionFilePlugin.js#L49) 🤔?
If the property already exists, we just skip reading the file again.

in the latest implementation I attached it to [request.tsconfigFileData](https://github.com/webpack/enhanced-resolve/pull/463/files#diff-b7aef0969ef4cdabf7bce8e5bc46e2e60a774daa2d5581cf5e6447eab2e0c345R311).